File: F#/LearnF#/Capstone3/Operations.fs
module Capstone3.Operations

open System
open Capstone3.Domain


let deposit amount account = { account with Balance = amount + account.Balance }

let withdraw amount account =
    if amount > account.Balance then
        printf "Transaction rejected\n"
        account
    else
        printf "Transaction approved\n"
        { account with Balance = account.Balance - amount }

let auditAs (operationName : string) (audit : Account -> Transaction -> unit) (operation : decimal -> Account -> Account) (amount : decimal) (account : Account) : Account =
    let newAccount = operation amount account
    let transaction : Transaction = { Timestamp = DateTime.Now; Operation = operationName; Amount = amount; Accepted = account.Balance <> newAccount.Balance; }
    audit newAccount transaction
    newAccount

let loadAccount (owner : string) (accountId : System.Guid) (transactions : Transaction list) =
    let processTransaction (account : Account) (transaction : Transaction) =
        if not transaction.Accepted then
            account
        else
            if transaction.Operation = "withdraw" then
                { account with Balance = account.Balance - transaction.Amount }
            elif transaction.Operation = "deposit" then
                { account with Balance = account.Balance + transaction.Amount }
            else
                failwith "Invalid command"
    let initialAccount = { AccountId = accountId; Owner = { Name = owner; }; Balance = 0M; }
    List.fold processTransaction initialAccount transactions

File: Haskell/Playground/LearnF#/Capstone3/src/Operations.hs
module Operations
     ( Command (..)
     , BankOperation (..)
     , char2Command
     , tryGetBankOperation
     , initializeAccount
     , deposit
     , withdraw
     ) where

import qualified Data.UUID as DU
import Data.Decimal
import Domain
import Transactions


data BankOperation = Deposit
                   | Withdraw
    deriving (Show, Eq)

data Command = BankCommand BankOperation
             | Exit
    deriving (Show, Eq)

char2Command :: Char -> Maybe Command
char2Command c
    | c == 'w'  = Just $ BankCommand Withdraw
    | c == 'd'  = Just $ BankCommand Deposit
    | c == 'x'  = Just Exit
    | otherwise = Nothing

tryGetBankOperation :: Command -> Maybe BankOperation
tryGetBankOperation (BankCommand op) = Just op
tryGetBankOperation _                = Nothing

initializeAccount :: String -> DU.UUID-> [Transaction] -> Account
initializeAccount owner guid transactions =
    Account guid (Customer owner) amount
    where
        amount = foldr f 0 transactions
        f (Transaction _ op amount accepted) accum =
            if accepted then
                case op of
                    "deposit"  -> accum + amount
                    "withdraw" -> accum - amount
            else
                accum

deposit :: Account -> Decimal -> Account
deposit (Account id owner balance) amount = Account id owner (balance + amount)

withdraw :: Account -> Decimal -> Account
withdraw (Account id owner balance) amount = Account id owner (balance - amount)
